src/Refactoring-Core/RBFindAndReplaceRefactoring.class.st

```diff
@@ -0,0 +1,192 @@
+	^ RBExtractMethodRefactoring new
+			setOption: #useExistingMethod 
+			toUse:  [ :ref :aSelector | 
+				true];
+			yourself 
+]
+
+{ #category : #initialization }
+RBFindAndReplaceRefactoring >> find: aSelector of: aClass inAllHierarchy: aBoolean [
+	class := self classObjectFor: aClass.
+	selector := aSelector.
+	replacesAllHierarchy := aBoolean.
+]
+
+{ #category : #accessing }
+RBFindAndReplaceRefactoring >> findAndReplaceOcurrencesIn: aMethod [ 
+	self findOcurrencesIn: aMethod
+]
+
+{ #category : #accessing }
+RBFindAndReplaceRefactoring >> findOcurrencesIn: aMethod [
+	|methodNode sourceCode flag |
+	flag := false.
+	methodNode := aMethod ast.
+	sourceCode := methodNode sourceCode.
+	(self nodesOf: methodNode) do: [ :each |
+            each first < each last
+                ifTrue: [ 
+	self matchNodes do: [ :matchNode | matchNode
+                      match: (RBPatternParser parseExpression: (sourceCode copyFrom: each first to: each last ))
+                      onSuccess: [ :map |
+	self extract: ((each first to: each last) -> map) 
+	of: aMethod.
+	flag := true. ]
+                      onFailure: [  ] .
+						flag ifTrue: [ self findOcurrencesIn: (aMethod modelClass methodFor: aMethod selector).
+							^ self]]]]. 
+	methodNode body nodesDo: [ :node |
+		self matchNodes do: [ :matchNode | matchNode
+                      match: node
+                      onSuccess: [ :map |
+	self extract: ((node start to: node stop) -> map) 
+	of: aMethod.
+	flag := true.]
+                      onFailure: [  ] .
+						flag ifTrue: [ self findOcurrencesIn: (aMethod modelClass methodFor: aMethod selector).
+							^ self]]].
+]
+
+{ #category : #accessing }
+RBFindAndReplaceRefactoring >> matchNodes [
+	^ matchNodes ifNil: [ |visitor node sourceCode|
+		visitor := RBMatchVisitor new.
+		node := self method ast copy .
+		node acceptVisitor: visitor.
+		sourceCode := self replaceArgumentsByPattern: node newSource .
+		sourceCode := sourceCode copyFrom: (self method ast body statements first start + visitor difference) to: sourceCode size.
+		matchNodes := OrderedCollection new.
+		matchNodes add: (RBPatternParser parseExpression: sourceCode).
+		node lastIsReturn ifTrue: [  node hasMultipleReturns 
+		ifFalse: [ sourceCode := sourceCode copyReplaceAll: '^' with:  ''.
+		matchNodes add: (RBPatternParser parseExpression: sourceCode) ] ].
+		matchNodes]
+]
+
+{ #category : #accessing }
+RBFindAndReplaceRefactoring >> method [
+	^ method ifNil: [ method := class methodFor: selector ]
+]
+
+{ #category : #accessing }
+RBFindAndReplaceRefactoring >> nodesOf: methodNode [ 
+	|visitor node|
+	visitor := RBCombinatorVisitor new.
+	node := methodNode copy .
+	node acceptVisitor: visitor.
+	^ visitor combinations 
+]
+
+{ #category : #preconditions }
+RBFindAndReplaceRefactoring >> preconditions [ 
+
+	| condition |
+	condition :=(RBCondition definesSelector: selector in: class) 
+		& (replacesAllHierarchy 
+		ifFalse: [ RBCondition empty ] 
```

Comment:
I would create a method emptyCondition as 
^ RBCondition empty
and avoid to hardocde all the RBCondition